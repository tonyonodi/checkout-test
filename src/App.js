import React, { useReducer } from "react";
import styled from "styled-components";
import Form from "./components/form";
import Comments from "./components/comments";
import "./App.css";
import isWhitespace from "./lib/isWhitespace";

const AppView = styled.div`
  width: 960px;
  max-width: 90%;
  margin: 0 auto;
`;

export const initialState = {
  formValues: {
    name: "",
    email: "",
    rating: undefined,
    comment: "",
  },
  formErrors: {},
  comments: [
    {
      name: "Alice",
      email: "alice@gmail.com",
      rating: 4,
      comment: "Some comment.",
      timeSubmitted: new Date(),
      id: 1,
    },
  ],
};

export const reducer = (state, action) => {
  switch (action.type) {
    case "update_form_value":
      return {
        ...state,
        formValues: {
          ...state.formValues,
          ...action.formUpdates,
        },
      };

    case "add_comment":
      let formErrors = {};
      if (isWhitespace(state.formValues.name)) {
        formErrors.name = "Name cannot be empty.";
      }
      if (isWhitespace(state.formValues.email)) {
        formErrors.email = "Email cannot be empty.";
      }
      if (!state.formValues.rating) {
        formErrors.rating = "Rating cannot be empty.";
      }
      if (Object.keys(formErrors).length > 0) {
        return {
          ...state,
          formErrors,
        };
      }

      return {
        ...state,
        formValues: initialState.formValues,
        formErrors: {},
        comments: [
          ...state.comments,
          { ...state.formValues, timeSubmitted: action.time, id: action.id },
        ],
      };

    default:
      return state;
  }
};

const App = () => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const updateFormValues = formUpdates =>
    dispatch({
      type: "update_form_value",
      formUpdates,
    });

  const addComment = () => {
    dispatch({ type: "add_comment", time: new Date(), id: Math.random() });
  };

  return (
    <AppView className="App">
      <Form
        values={state.formValues}
        errors={state.formErrors}
        updateValues={updateFormValues}
        addComment={addComment}
      />
      <Comments comments={state.comments} />
    </AppView>
  );
};

export default App;
