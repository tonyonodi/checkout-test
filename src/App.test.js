import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import App, { initialState, reducer } from "./App";

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it("renders correctly", () => {
  const tree = renderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe("app reducer", () => {
  test("can update form values", () => {
    const dispatch = {
      type: "update_form_value",
      formUpdates: { name: "Alice" },
    };

    const firstUpdate = reducer(initialState, dispatch);

    expect(firstUpdate).toEqual({
      formValues: {
        name: "Alice",
        email: "",
        rating: undefined,
        comment: "",
      },
      formErrors: {},
      comments: initialState.comments,
    });
  });

  test("can add comment from correct form values", () => {
    const dispatch = { type: "add_comment", time: new Date(), id: 1 };
    const state = {
      formValues: {
        name: "Bob",
        email: "bob@gmail.com",
        rating: 3,
        comment: "Another comment.",
      },
      formErrors: {},
      comments: [],
    };

    const firstUpdate = reducer(state, dispatch);

    const firstExpected = {
      formValues: {
        name: "",
        email: "",
        rating: undefined,
        comment: "",
      },
      formErrors: {},
      comments: [
        {
          name: "Bob",
          email: "bob@gmail.com",
          rating: 3,
          comment: "Another comment.",
          timeSubmitted: dispatch.time,
          id: 1,
        },
      ],
    };

    expect(firstUpdate).toEqual(firstExpected);
  });

  test("shows errors for missing values", () => {
    const dispatch = { type: "add_comment", time: new Date(), id: 1 };
    const state = {
      formValues: {
        name: "",
        email: "",
        rating: undefined,
        comment: "",
      },
      formErrors: {},
      comments: [],
    };

    const firstUpdate = reducer(state, dispatch);

    const firstExpected = {
      formValues: {
        name: "",
        email: "",
        rating: undefined,
        comment: "",
      },
      formErrors: {
        name: "Name cannot be empty.",
        email: "Email cannot be empty.",
        rating: "Rating cannot be empty.",
      },
      comments: [],
    };

    expect(firstUpdate).toEqual(firstExpected);
  });
});
