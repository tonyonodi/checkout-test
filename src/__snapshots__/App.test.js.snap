// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="App sc-gZMcBi fiETDR"
>
  <form
    className="sc-bdVaJa bPGLpq"
    onSubmit={[Function]}
  >
    <div>
      <label
        className="sc-bwzfXH uTxCW"
      >
        Name *
      </label>
      <input
        className="sc-htpNat drnJZU"
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        className="sc-bwzfXH uTxCW"
      >
        Email *
      </label>
      <input
        className="sc-htpNat drnJZU"
        name="email"
        onChange={[Function]}
        type="email"
        value=""
      />
    </div>
    <div>
      <label
        className="sc-bwzfXH uTxCW"
      >
        Rating *
      </label>
      <input
        className="sc-htpNat drnJZU"
        max="5"
        min="0"
        name="rating"
        onChange={[Function]}
        step="1"
        type="number"
      />
    </div>
    <div />
    <label
      className="sc-bwzfXH uTxCW"
    >
      Comment
    </label>
    <textarea
      className="sc-bxivhb dqNOXg"
      name="comment"
      onChange={[Function]}
      value=""
    />
    <input
      className="sc-htpNat drnJZU"
      type="submit"
    />
  </form>
  <div
    className="sc-dnqmqq eDLMuH"
  >
    There are no comments yet.
  </div>
</div>
`;
